### PR TITLE
Improve code navigation for unreferenced declaration

### PR DESCRIPTION
**Pull request by keyboardDrummer**
### Description
- Renamed `IHasUsages` to `IHasReferences`, to be more in line with LSP naming
- Renamed `IDeclarationOrUsage` to `IHasNavigationToken`, and its member `NameToken` to `NavigationToken`
- Respect the `request.Context.IncludeDeclaration` value of the LSP references request
- Enable renaming declarations that were not referred to

### How has this been tested?
- Added tests for renaming and finding references on a declaration that is not referred to

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyCore/AST/TypeDeclarations/TopLevelDecl.cs
Comment (atomb) [inline Source/DafnyCore/AST/TypeDeclarations/TopLevelDecl.cs:4]: I'm a little sad to have the `DafnyCore` package depend on LSP-specific stuff, but I think the tradeoff is probably worth it.

File: Source/DafnyCore/AST/Types/TypeParameter.cs
Comment (atomb) [inline Source/DafnyCore/AST/Types/TypeParameter.cs:203]: I'm not sure how to interpret this comment. Have you tested calling `GetDescription` on a `TypeParameter`?

File: Source/DafnyCore/AST/TypeDeclarations/TopLevelDecl.cs
Comment (atomb) [reply Source/DafnyCore/AST/TypeDeclarations/TopLevelDecl.cs:4]: Looking more closely, I see that isn't a new dependency, so it's even more on the side of being worth it.

**Review by atomb — APPROVED** [review]
I think this looks good. There's one comment that should probably be either removed or addressed, but I don't think that's worth blocking the merge over. I'm happy to re-approve if you do make changes related to that.

File: Source/DafnyCore/AST/TypeDeclarations/TopLevelDecl.cs
Comment (keyboardDrummer) [reply Source/DafnyCore/AST/TypeDeclarations/TopLevelDecl.cs:4]: I think it's to be expected that a compiler will have a public interface so it can be used as a library, and LSP happens to already contain many types that are useful for this interface, such as `Position` and `Range`. I've made this PR to reduce what's in scope from DafnyCore: https://github.com/dafny-lang/dafny/pull/5502

**Pull request merged**
